File: src/cli/help.test.ts
/** @fileoverview Exercises command discovery and usage errors through the real CLI without relying on help snapshots. */

import { test, expect } from 'bun:test';
import { mkdtemp, readdir, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';

const entry = fileURLToPath(new URL('../cli.ts', import.meta.url));

async function invoke(
  args: readonly string[],
): Promise<ReturnType<typeof spawnSync>> {
  const directory = await mkdtemp(join(tmpdir(), 'code-rules-help-'));
  try {
    const result = spawnSync(process.execPath, [entry, ...args], {
      cwd: directory,
      encoding: 'utf8',
      timeout: 10000,
    });
    expect(await readdir(directory)).toEqual([]);
    return result;
  } finally {
    await rm(directory, { recursive: true, force: true });
  }
}

test('top-level help supports discovery without dumping authoring options', async () => {
  const result = await invoke(['--help']);
  expect(result.status).toBe(0);
  expect(result.stderr).toBe('');
  expect(result.stdout).toContain(
    'The package manager for your engineering rules',
  );
  expect(result.stdout).toContain('Project commands:');
  expect(result.stdout).toContain('add source');
  expect(result.stdout).toContain('Library commands:');
  expect(result.stdout).not.toContain('--impact-description');
  expect(result.stdout).not.toContain('--license-file');
});

test('bare command groups show navigation without requiring project setup', async () => {
  for (const args of [[], ['local'], ['library'], ['library', 'add']]) {
    const result = await invoke(args);
    expect(result.status).toBe(0);
    expect(result.stdout).toContain('<command>');
    expect(result.stdout).toContain('--help');
  }
});

test('nested help exposes only applicable options, defaults, and examples', async () => {
  const library = await invoke(['library', 'init', '--help']);
  expect(library.status).toBe(0);
  expect(library.stdout).toContain('code-rules library init [options]');
  expect(library.stdout).toContain('--license-file');
  expect(library.stdout).toContain('default: current directory');
  expect(library.stdout).toContain('without prompts');
  expect(library.stdout).toContain('Examples:');
  expect(library.stdout).not.toContain('--config');
  const local = await invoke(['local', 'add', 'rule', '-h']);
  expect(local.status).toBe(0);
  expect(local.stdout).toContain('--body-file');
  expect(local.stdout).toContain('.code-rules/config.json');
  expect(local.stdout).toContain('--non-interactive');
  expect(local.stdout).not.toContain('--license-file');
});

test.each([
  {
    args: ['library', 'unknown'],
    message: 'Unknown command "unknown"',
    hint: 'code-rules library --help',
  },
  {
    args: ['init', '--directory', 'elsewhere'],
    message: 'Unknown option --directory',
    hint: 'code-rules init --help',
  },
  {
    args: ['library', 'init', '--config', 'elsewhere'],
    message: 'Unknown option --config',
    hint: 'code-rules library init --help',
  },
  {
    args: ['local', 'add', 'rule'],
    message: 'Missing <rule-id>',
    hint: 'code-rules local add rule --help',
  },
  {
    args: ['check', '--config'],
    message: 'Provide a value for --config',
    hint: 'code-rules check --help',
  },
  {
    args: ['check', '--config', 'one', '--config', 'two'],
    message: 'cannot repeat',
    hint: 'code-rules check --help',
  },
  {
    args: ['build', 'extra', '--help'],
    message: 'Unexpected argument "extra"',
    hint: 'code-rules build --help',
  },
])(
  'usage failure points to the relevant help: $args',
  async ({ args, message, hint }) => {
    const result = await invoke(args);
    expect(result.status).toBe(2);
    expect(result.stdout).toBe('');
    expect(result.stderr).toContain(message);
    expect(result.stderr).toContain(hint);
  },
);

test('requesting help bypasses authoring prompts and reads even with nonexistent input paths', async () => {
  const result = await invoke([
    'library',
    'init',
    '--license-file',
    'missing',
    '--help',
  ]);
  expect(result.status).toBe(0);
  expect(result.stdout).toContain('--spdx');
});

test('missing noninteractive metadata identifies the field and the command help', async () => {
  const result = await invoke([
    'local',
    'add',
    'group',
    'practices/testing',
    '--non-interactive',
  ]);
  expect(result.status).toBe(2);
  expect(result.stderr).toContain('--name');
  expect(result.stderr).toContain('code-rules local add group --help');
});

test('help stays readable at 80 columns without color or broken quoted examples', async () => {
  const result = await invoke(['local', 'add', 'rule', '--help']);
  expect(result.status).toBe(0);
  const output = String(result.stdout);
  expect(output).not.toContain('\u001b[');
  expect(
    Math.max(...output.split('\n').map((line) => line.length)),
  ).toBeLessThanOrEqual(80);
  const example = output.split('Examples:\n')[1]?.replace(/\\\n\s+/gu, '');
  expect(example).toContain('--title "Limit retries"');
  expect(example).toContain('--when-to-read "When changing retries"');
});

File: internal/cli/help_test.go
// Check option grouping through real commands, including flags whose meaning depends on scope.

package cli

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

func TestCommandOptionSections(t *testing.T) {
	binary := buildCLI(t)
	directory := t.TempDir()
	for _, test := range []struct {
		path, title      string
		specific, common []string
	}{
		{"", "", nil, []string{"help", "json", "version"}},
		{"project", "", nil, []string{"help", "json"}},
		{"project add", "", nil, []string{"help", "json"}},
		{"project init", "", nil, []string{"config", "help", "json", "non-interactive"}},
		{"project sync", "", nil, []string{"config", "help", "json"}},
		{"project build", "", nil, []string{"config", "help", "json"}},
		{"project check", "", nil, []string{"config", "help", "json"}},
		{"project add library", "Library options", []string{"repository", "ref", "groups"}, []string{"config", "help", "json", "non-interactive"}},
		{"project add rule", "Rule options", []string{"title", "when-to-read", "impact", "impact-description", "body-file"}, []string{"config", "help", "json", "non-interactive"}},
		{"library", "", nil, []string{"help", "json"}},
		{"library add", "", nil, []string{"help", "json"}},
		{"library init", "Library options", []string{"spdx", "license-file", "notice-file"}, []string{"directory", "help", "json", "non-interactive"}},
		{"library check", "", nil, []string{"directory", "help", "json", "non-interactive"}},
		{"library add rule", "Rule options", []string{"title", "when-to-read", "impact", "impact-description", "body-file"}, []string{"directory", "help", "json", "non-interactive"}},
	} {
		t.Run(test.path, func(t *testing.T) {
			for _, structured := range []bool{false, true} {
				args := append(strings.Fields(test.path), "--help")
				if structured {
					args = append([]string{"--json"}, args...)
				}
				out, diagnostic, code := runCLI(t, binary, directory, args...)
				if code != 0 || diagnostic != "" {
					t.Fatal(args, code, out, diagnostic)
				}
				if structured {
					var result struct {
						OK    bool
						Value struct{ Text string }
					}
					if err := json.Unmarshal([]byte(out), &result); err != nil || !result.OK {
						t.Fatal(err, out)
					}
					out = result.Value.Text
				}
				before, common, found := strings.Cut(out, "Common options:\n")
				if !found || strings.Contains(out, "Flags:") {
					t.Fatal("missing common options", out)
				}
				// Examples may repeat flags after the options section.
				common, _, _ = strings.Cut(common, "\n\n")
				specific := ""
				if test.title != "" {
					_, specific, found = strings.Cut(before, test.title+":\n")
					if !found {
						t.Fatal("missing command-specific options", out)
					}
				} else if strings.Contains(before, " options:") {
					t.Fatal("empty options section", out)
				}
				for _, flag := range test.specific {
					if strings.Count(specific, "--"+flag+" ") != 1 || strings.Contains(common, "--"+flag+" ") {
						t.Fatal("wrong specific option section", flag, out)
					}
				}
				for _, flag := range test.common {
					if strings.Count(common, "--"+flag+" ") != 1 || strings.Contains(specific, "--"+flag+" ") {
						t.Fatal("wrong common option section", flag, out)
					}
				}
				if strings.Contains(common, "--config ") && strings.Contains(common, "--directory ") {
					t.Fatal("options from another scope leaked", out)
				}
			}
		})
	}
	entries, err := os.ReadDir(directory)
	if err != nil || len(entries) != 0 {
		t.Fatal("help changed files", entries, err)
	}
}
